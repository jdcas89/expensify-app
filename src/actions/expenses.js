import uuid from "uuid";
import * as types from './actionTypes';
import database from '../firebase/firebase';




export const addExpense = (expense) => ({ 
  type: types.ADD_EXPENSE, 
  expense
});

export const removeExpense = ({ id }) => ({
  type: types.REMOVE_EXPENSE,
  id
});

export const editExpense = (id, updates) => ({
  type: types.EDIT_EXPENSE,
  id,
  updates
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "", note = "", amount = 0, createdAt = 0
    } = expenseData;
    const expense = {description, note, amount , createdAt}
    database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}