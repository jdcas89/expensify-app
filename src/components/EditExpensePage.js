import React from "react";

const EditExpensePage = (props) => {
  console.log(props)
  return (
    <div>
      <h2>Edit Page</h2>
      <p>Editing the page of item {props.match.params.id}</p>
    </div>
  );
};

export default EditExpensePage;
