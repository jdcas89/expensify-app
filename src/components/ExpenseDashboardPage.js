import React from 'react';

const ExpenseDashboardPage = () => (
  <div>
    <h2>Home Page</h2>
  </div>
);

export default ExpenseDashboardPage;