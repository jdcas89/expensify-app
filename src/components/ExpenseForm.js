import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      errorMessage: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    e.persist();
    this.setState(() => ({ note: e.target.value }));
  };

  // https://regex101.com/
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        errorMessage: "You have to provide a description and an amount!"
      }));
    } else {
      this.setState(() => ({ errorMessage: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.errorMessage && <p>{this.state.errorMessage}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            placeholder="Description"
            autoFocus
          />
          <input
            type="text"
            value={this.state.amount}
            onChange={this.onAmountChange}
            placeholder="Amount"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder="Add a note for your expense"
          />

          <button>Add expense</button>
        </form>
      </div>
    );
  }
}
