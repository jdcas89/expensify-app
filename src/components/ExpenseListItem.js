import React from "react";
import { Link } from 'react-router-dom'
import moment from 'moment'

const ExpenseListItem = ({ id, description, note, amount, createdAt }) => (
  <div>
  <Link to={`/edit/${id}`}>  <h3>{description}</h3></Link>
   
    <p> <strong>Notes: </strong> {note}</p>
    <p> <strong>Amount: </strong> {amount}</p>
    <p><strong> Date created:</strong> {moment(createdAt).format('MMMM Do YYYY')}</p>
    <hr />
  </div>
);

export default ExpenseListItem;