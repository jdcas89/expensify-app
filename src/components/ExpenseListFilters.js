import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";


class ExpenseListFilters extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      calendarFocused: null
    }
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({calendarFocused}))
  }
  onDatesChange = ({startDate, endDate}) => {
    this.props.dispatch(setStartDate(startDate))
    this.props.dispatch(setEndDate(endDate))
  }
  render(){
    return (
      <div>
        <p>Filter text: </p>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => {
            props.dispatch(setTextFilter(e.target.value));
          }}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={(e) => {
            console.log(e.target.value)
            if(e.target.value === 'date'){
              this.props.dispatch(sortByDate());
            }else if(e.target.value === 'amount'){
              this.props.dispatch(sortByAmount());
            }
          }}
          >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
        startDate={this.props.filters.startDate} 
        endDate={this.props.filters.endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        numberOfMonths={1}
        isOutsideRange= {() => false}
        showClearDates={true}
      />      
      </div>
    )
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
