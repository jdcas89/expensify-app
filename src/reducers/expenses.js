import * as types from '../actions/actionTypes'
//Expenses reducer

const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case types.ADD_EXPENSE:
      return [...state, action.expense]
    case types.EDIT_EXPENSE:
      return state.map((expense) => {
        if(expense.id === action.id){
          return {
            ...expense,
            ...action.updates
          };
        }else{
          return expense;
        }
      })
    case types.REMOVE_EXPENSE:
      return state.filter(({ id }) => id !== action.id);
    case types.SET_EXPENSES:
      return action.expenses;
    default:
      return state;
  }
};

//export default expensesReducer;