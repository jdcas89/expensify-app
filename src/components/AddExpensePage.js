import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
  addExpense = expense => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h2>Create Expense</h2>
        <ExpenseForm onSubmit={this.addExpense} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
