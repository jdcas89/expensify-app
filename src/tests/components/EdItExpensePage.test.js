//refactor editpage to class component
//setup mapDispatchToProps
// should render EditExpensePage
//Should handle edit expense spies
//Should handle remove expense spies

import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      expense = {expenses[0]}
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
    />
  );
});

test("should render EditExpensePage", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle on submit", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test("should handle remove expense", () => {
  wrapper.find("button").simulate('click')
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(removeExpense).toHaveBeenLastCalledWith({
    id:expenses[0].id
  });
});