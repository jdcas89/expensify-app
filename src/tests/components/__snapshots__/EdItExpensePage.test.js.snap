// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <h2>
    Edit Page
  </h2>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "Biology",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Delete expense
  </button>
</div>
`;
